### ISSUE TITLE
"Boxed" barcodes have a poor detection rate

### ISSUE THREAD

**Issue by sascha-hendel** [serratus/quaggaJS#106]
EAN barcodes on colored book covers are often printed in a small box with white background. In this case the detection rate is often pretty worse. If the whole background of the book cover including barcode is in the same color, detection is much better.
Maybe Quagga algorithm thinks, that the edge(s) of the box are part of the barcode ?

**Comment by serratus** [comment]
This problem is most probably caused by the binarization step. The algorithm tries to separate the background from the foreground globally. Changing this to a local threshold could improve the recognition rate significantly. I've played around with different methods for binarization, but the current implementation proved to be the best fit in the common case.

For your specific use-case, especially if the barcodes are really small compared to the colored background, a local threshold would be a better fit. I might find a better solution in the future, but currently I don't have much time to research on new ideas. Sorry for that.

**Comment by sascha-hendel** [comment]
Does images are converted in Quagga into grayscale before processing? If yes, I would like to suggest a different way:
Nearly all barcodes are printed in black or grey (I haven't seen any colored barcodes, yet). Out of this reason, it could be better to convert all colored areas into plain white and not into different grayscales. This would equalize all unnecessary edges between different colored areas and simplify further processing.

**Comment by serratus** [comment]
Good point. I haven't thought of that optimization yet. I woulds say that covers around 99% of use-cases. I have seen a few colored barcodes, but one might configure it explicitly to allow colored barcodes. Let me experiment with that and I'll let you know what I discovered.

**Comment by ericblade** [comment]
I wonder if @sascha-hendel or @nashiralyana might have some sample images . I've seen a few that have had a lot of difficulty getting, and I should probably take some images to build some test suite enhancements with.

**Comment by sascha-hendel** [comment]
Sample "boxed" barcodes should be found on most book covers.
I bought in the past new hardware (new tablet and smartphone with better camera) and validate recognized isbn codes before processing. Now boxed barcodes aren't a problem for me anymore and therefore I can't find "problematic" samples to provide for testing, sorry!

**Comment by ericblade** [comment]
Interesting.  I hadn't really thought about it much, and I haven't done a lot of book scanning recently, but I was thinking that colored backgrounds on the barcodes were much more difficult to deal with.  Maybe I need to start taking pictures of barcodes that don't seem to scan very well, to add them to the test suites. :-)

Glad to see you're still using this after 2 years, especially after almost a year of the project being idle. :)

I'm thinking in the future (maybe not too distant :) ) we might end up writing some documentation, and/or some code for improving the scan accuracy for various types of codes, if you'd like to either contribute some code for your isbn validation, or some description of the process, if you don't want to or can't contribute the code.

**Comment by sascha-hendel** [comment]
I use Quagga.js in an online book database ( www.book-info.com ). You can easily try it by yourself. If your hardware and browser are suitable, a book scan option shows up in the navigation menue. Feedback & user experience is always welcome!

**Comment by ericblade** [comment]
I can see it, though quagga can't get any decent images through my PC camera to check out it's accuracy.. i'll have to load it from a mobile after i get some battery power.  I don't see where the code for validating the result from Quagga might be :(

**Comment by sascha-hendel** [comment]
Same in my case: My laptop (older model) camera does recognize barcodes only in perfect circumstances. But with tablet and smartphone (newer stuff from Samsung) barcode recognition works pretty well.
ISBN validation is done in Javascript in the ondetect method.

**Comment by ericblade** [comment]
Interestingly, I had to look it up using Edge to actually find the source code in the .htm file. Chrome and Vivaldi were failing to give me any visual on it.  

I'm amazed at how well mobile cameras seem to work compared to PC cameras.  

Thanks for the help.
